phantom/telemetry.py

```diff
@@ -0,0 +1,393 @@
+
+    def log_resolver_round(self, round: int, max: Optional[int]) -> None:
+        if self._enable_print and self._print_messages:
+            print(
+                _t(1)
+                + colored(
+                    f"Batch Resolver round {round+1}/{max or 'Inf'}:",
+                    color="grey",
+                )
+            )
+
+    def log_msg_send(self, message: Message) -> None:
+        self._print_msg(message, "SEND")
+
+    def log_msg_recv(self, message: Message) -> None:
+        if self._current_episode is not None:
+            self._current_episode["steps"][-1]["messages"].append(asdict(message))
+
+        self._print_msg(message, "RECV")
+
+    def log_metrics(self, env: "PhantomEnv") -> None:
+        if self._current_episode is not None:
+            self._current_episode["steps"][-1]["metrics"] = {
+                name: metric.extract(env) for name, metric in self._file_metrics.items()
+            }
+
+        if self._enable_print and self._print_metrics is not None:
+            print(_t(1) + colored("METRICS:", color="cyan"))
+
+            if len(self._print_metrics) > 0:
+                for name, metric in self._print_metrics.items():
+                    print(_t(2) + f"{name: <30} : {metric.extract(env)}")
+            else:
+                print(_t(2) + "None")
+
+    def log_episode_done(self) -> None:
+        self._write_episode_to_file()
+
+        if self._enable_print:
+            print(_t(1) + colored("EPISODE DONE", color="green", attrs=["bold"]))
+
+    def _print_msg(self, message: Message, string: str) -> None:
+        if self._should_print_msg(message):
+            route_str = f"{message.sender_id: >10} --> {message.receiver_id: <10}"
+            msg_name = f"({message.payload.__class__.__name__})"
+            fields = ", ".join(f"{k}: {v}" for k, v in message.payload.__dict__.items())
+
+            print(
+                _t(2)
+                + f"MSG {string}: {route_str} {msg_name: <20}"
+                + colored(fields, attrs=["dark"])
+            )
+
+    def _write_episode_to_file(self) -> None:
+        if self._output_file is not None and self._current_episode is not None:
+            json.dump(
+                self._current_episode,
+                self._output_file,
+                indent=2 if self._human_readable else None,
+                cls=NumpyArrayEncoder,
+            )
+            self._output_file.write("\n")
+            self._output_file.flush()
+            self._current_episode = None
+
+    def _should_print_msg(self, message: Message) -> bool:
+        return (
+            self._enable_print
+            and self._print_messages
+            and (
+                isinstance(self._print_messages, bool)
+                or message.sender_id in self._print_messages
+                or message.receiver_id in self._print_messages
+            )
+        )
+
+
+def _t(n: int) -> str:
```

Comment:
Maybe use string formatting instead of your own function:
https://stackoverflow.com/a/5676884

Reply:
I can replace the function calls with:
```python
f"{' ':{2*TAB_SIZE}}"
```
and merge any contiguous strings but IMO it makes it less clean.

Reply:
ok, fair enough